File: frontend/src/services/rating.service.js
import {api} from "../api.js";

export const addOrUpdateRating = async (data) => {
    try {
        const response = await api.post("/ratings", data);
        return response.data;
    } catch (e) {
        console.error(e);
    }
};

export const getUserRatingForFood = async (userId, foodId) => {
    try {
        const response = await api.get(`/ratings/user/${userId}/food/${foodId}`);
        return response.data;
    } catch (e) {
        console.error(e);
    }
};

export const getUserRatings = async (userId) => {
    try {
        const response = await api.get(`/ratings/user/${userId}`);
        return response.data;
    } catch (e) {
        console.error(e);
    }
};

export const getAllRatingsForFood = async (foodId) => {
    try {
        const response = await api.get(`/ratings/food/${foodId}`);
        return response.data;
    } catch (e) {
        console.error(e);
    }
};

export const getFoodAverageRating = async (foodId) => {
    try {
        const response = await api.get(`/ratings/food/${foodId}/average`);
        return response.data;
    } catch (e) {
        console.error(e);
    }
};

export const deleteRating = async (userId, foodId) => {
    try {
        const response = await api.delete(`/ratings/user/${userId}/food/${foodId}`);
        return response.data;
    } catch (e) {
        console.error(e);
    }
};

export const getAllRatings = async () => {
    try {
        const response = await api.get("/ratings");
        return response.data;
    } catch (e) {
        console.error(e);
    }
};

File: backend/src/main/java/com/sliit/crave/service/RatingService.java
package com.sliit.crave.service;

import com.sliit.crave.dto.request.rating.RatingRequest;
import com.sliit.crave.dto.response.Response;
import com.sliit.crave.dto.response.rating.AverageRatingResponse;
import com.sliit.crave.dto.response.rating.RatingResponse;

import java.util.List;

public interface RatingService {
    
    // Add or update rating (upsert)
    Response<RatingResponse> addOrUpdateRating(RatingRequest request);
    
    // Get user's rating for a specific food
    Response<RatingResponse> getUserRatingForFood(Long userId, Long foodId);
    
    // Get all ratings by a user
    Response<List<RatingResponse>> getUserRatings(Long userId);
    
    // Delete user's rating for a food
    Response<Void> deleteRating(Long userId, Long foodId);
    
    // Get all ratings for a food (Admin)
    Response<List<RatingResponse>> getAllRatingsForFood(Long foodId);
    
    // Get average rating and total count for a food (Public)
    Response<AverageRatingResponse> getAverageRatingResponse(Long foodId);
    
    // Get all ratings (Admin)
    Response<List<RatingResponse>> getAllRatings();

    // Get average rating for a food
    Double getAverageRating(Long foodId);

    // Get total ratings count for a food
    Long getTotalRatings(Long foodId);
}
